xprocess.py

```diff
@@ -166,32 +182,37 @@ def ensure(self, name, preparefunc, restart=False):
             else:
                 kwargs["close_fds"] = True
                 kwargs["preexec_fn"] = os.setpgrp  # no CONTROL-C
-            self.running_procs.append(
-                Popen(
-                    args,
-                    **popen_kwargs,
-                    **kwargs,
-                )
-            )
-            info.pid = pid = self.running_procs[-1].pid
+
+            # keep references of all popen
+            # and info objects for cleanup
+            self._info_objects.append((info, starter.terminate_on_interrupt))
```

Comment:
follow-up idea - lets encapsulate the info, popen and log into a object that does resoruce cleanp
the co-located lists scream scared concern

Reply:
I agree it is a good idea, and up to @northernSage to follow up. 😁 

If @northernSage decides to follow up however, I suggest doing so in a separate PR for easier reviewing. 👍

Reply:
That seems like a good idea to me as well. Will open a issue to document this for now and get started on the follow up PR soon 🚀

Thanks for the help @RonnyPfannschmidt  and @nicoddemus